### PR TITLE
dotnet core cli build

### PR DESCRIPTION
**Pull request by ncave**
Rough cut, work in progress, just for evaluation and comments (maybe put it in a branch).
FCS builds successfully, but:
- no xplat tools yet, so all generated files are checked in an output folder
- tests not fixed yet
- no appveyor / travis yet

### CODE REVIEW

**Comment by ncave** [comment]
@dsyme Started a first rough cut [here](https://github.com/fsharp/FSharp.Compiler.Service/pull/556).

**Comment by ncave** [comment]
This PR is a follow-up of [this discussion](https://github.com/fsharp/FSharp.Compiler.Service/issues/465).

**Comment by ncave** [comment]
@dsyme I can probably use some advice on how to proceed:

FCS builds successfully (netcore only), and I can work on the tests / adding appveyor / travis etc. but I'm not sure if I'm going in the right direction, so I have to stop and ask. A lot of my confusion is due to my limited understanding of the code base evolution and timeline in relation to upstream.

I made it work by mostly clobbering your [#555](https://github.com/fsharp/FSharp.Compiler.Service/pull/555) with ms/visualfsharp code plus some adaptations here and there. I'm not sure why is there such a big difference between ms/visualfsharp and fsharp/fsharp, are they evolving completely separately?

Can you please spot check a few files in the PR and tell me if I lost too many important changes in the process? A lot of the difference is mono-related, and the code is already riddled with ifdefs as it is, so adding more ifdefs will probably make it less maintainable in the long run.

If I need to retain more of the [#555](https://github.com/fsharp/FSharp.Compiler.Service/pull/555) code, I'll have to scrap this PR and start over. I may be going about it entirely the wrong way, are there any major coreclr PRs in ms/visualfsharp that I can just merge on top of 555 in FCS?

Thank you for your help!

File: src/absil/il.fs [hunk 1]
```diff
@@ -13,22 +13,17 @@ open Microsoft.FSharp.Compiler.AbstractIL
...
 open System.Collections.Concurrent
  
 let logging = false 
 
-let runningOnWindows =
-    match System.Environment.OSVersion.Platform with
-    | PlatformID.Win32NT | PlatformID.Win32S | PlatformID.Win32Windows | PlatformID.WinCE -> true
-    | _  -> false
-
+let runningOnMono =
```
Comment (dsyme) [inline src/absil/il.fs]: I don't yet understand why these changes were not integrated across from Microsoft\visualfsharp in the latest integration.  I will look into that.  You shouldn't need to be bringing them across - they should just be there

**Comment by dsyme** [comment]
@ncave OK, for some reason not all the CoreCLR compiler changes had yet been integrated across.  I'm doing that now, first in [fsharp/fsharp](https://github.com/fsharp/fsharp/pull/567).  This PR should get much, much smaller after a complete integration

**Comment by ncave** [comment]
Great, it looks like [#567](https://github.com/fsharp/fsharp/pull/567) should solve my tools issue as well. I'll wait for it then.

**Comment by dsyme** [comment]
ok @ncave, https://github.com/fsharp/FSharp.Compiler.Service/pull/557 has been integrated, which should help

**Comment by ncave** [comment]
@dsyme Thanks, that helped a lot to reduce the difference. I got it to build again (the netcore FCS), but there are still quite a few differences in fsc.fs and fsi.fs between ms/visualfsharp and here, if you want to take a look just in case.
Still work to do:
- dotnet-fssrgen not used yet (outputs kept in the output folder temporarily, pls ignore)
- tests / appveyor / travis not started yet

File: src/absil/il.fs [hunk 2]
```diff
@@ -13,18 +13,12 @@ open Microsoft.FSharp.Compiler.AbstractIL
...
 open Microsoft.FSharp.Compiler.AbstractIL.Internal.Library
 open Microsoft.FSharp.Compiler.AbstractIL.Diagnostics
-open System
 open System.Collections
 open System.Collections.Generic
 open System.Collections.Concurrent
  
 let logging = false 
 
-let runningOnWindows =
```
Comment (dsyme) [inline src/absil/il.fs]: Keep this code please

File: src/absil/il.fsi [hunk 3]
```diff
@@ -2276,8 +2276,9 @@ type ILPropertyRef =
      member Name: string
      interface System.IComparable
 
-val runningOnWindows: bool
+#if ENABLE_MONO_SUPPORT
```
Comment (dsyme) [inline src/absil/il.fsi]: remove this thanks (prefer what is in the FCS repository please)

File: src/absil/ilread.fs [hunk 4]
```diff
@@ -3970,15 +3970,16 @@ let ClosePdbReader pdb =
 #endif
 
 let OpenILModuleReader infile opts = 
-    if IL.runningOnWindows then
```
Comment (dsyme) [inline src/absil/ilread.fs]: remove the changes in this file thanks (prefer what is in the FCS repository please)

File: src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.fs [hunk 5]
```diff
@@ -0,0 +1,329 @@
+namespace FSIstrings
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.fs]: Don't commit this file or the other two files just above please

File: src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.resx [hunk 6]
```diff
@@ -0,0 +1,276 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (dsyme) [inline src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.resx]: Don't commit this file please

File: src/fsharp/fsi/fsi.fs [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
-module Microsoft.FSharp.Compiler.Interactive.Shell
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: In general you should undo all the changes in this file. Some additional changes _may_ be needed to get it to compile on .NET Core. However this file has largely diverged from Microsoft\visualfsharp (or at least the changes here have not yet been folded back into that repo).  So we effectively have to work out how to compile this on .NET Core.  Alternatively remove it from the .netcore version of FSharp.Compiler.Service to start, depending on whether you need dynamic compilation or not.

Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: Actually looking at what you have it seems you may be close to having done this as I described (sorry).

File: src/fsharp/fsi/fsi.fs [hunk 8]
```diff
@@ -89,6 +93,7 @@ module internal Utilities =
 
     let ignoreAllErrors f = try f() with _ -> ()
 
+let referencedAssemblies = Dictionary<string, DateTime>()
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: I'm not sure why Microsoft\visualfsharp has moved this out to be a global.  It looks wrong to do this for FCS.

File: src/fsharp/fsi/fsi.fs [hunk 9]
```diff
@@ -924,7 +937,11 @@ type internal FsiDynamicCompiler
         errorLogger.AbortOnError(fsiConsoleOutput);
             
         let fragName = textOfLid prefixPath 
+#if ENABLE_MONO_SUPPORT
         let codegenResults = GenerateIlxCode (IlReflectBackend, isInteractiveItExpr, runningOnMono, tcConfig, topCustomAttrs, optimizedImpls, fragName, true, ilxGenerator)
+#else
+        let codegenResults = GenerateIlxCode (IlReflectBackend, isInteractiveItExpr, false, tcConfig, topCustomAttrs, optimizedImpls, fragName, true, ilxGenerator)
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: You don't need this change (in FCS repo runningOnMono is false when ENABLE_MONO_SUPPORT is not set)

File: src/fsharp/fsi/fsi.fs [hunk 10]
```diff
@@ -1329,7 +1346,7 @@ type internal FsiInterruptController(fsiOptions : FsiCommandLineOptions,
         ignore fsiConsoleOutput
         ignore CTRL_C
         ignore fsiOptions
-        false
+        exitViaKillThread <- false
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: We should work out what this change is for (was it a mistake in the FCS integration?)

File: src/fsharp/fsi/fsi.fs [hunk 11]
```diff
@@ -1499,17 +1518,21 @@ module internal MagicAssemblyResolution =
         Assembly.LoadFrom(path)
 #endif
 
-    let ResolveAssembly(m,tcConfigB, tcImports: TcImports, fsiDynamicCompiler: FsiDynamicCompiler, fsiConsoleOutput: FsiConsoleOutput, fullAssemName:string) = 
+    let Install(tcConfigB, tcImports: TcImports, fsiDynamicCompiler: FsiDynamicCompiler, fsiConsoleOutput: FsiConsoleOutput) =
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: I'm not sure what this change is doing.  I suspect the FCS version is the right one

File: src/fsharp/fsi/fsi.fs [hunk 12]
```diff
@@ -2789,9 +2821,9 @@ module Settings =
                      setSignal exitSignal
         interface System.IDisposable with 
              member x.Dispose() =
-                         runSignal.Close();
-                         exitSignal.Close();
-                         doneSignal.Close();
+                 runSignal.Dispose();
+                 exitSignal.Dispose();
+                 doneSignal.Dispose();
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: Not sure what these indentation changes are for - is it aligning code with Microsoft\visualfsharp? If so that's good.

File: src/fsharp/fsi/fsi.fsi [hunk 13]
```diff
@@ -81,11 +81,11 @@ type public FsiEvaluationSessionHostConfig =
     /// In this case, a prompt is printed early, a background thread is created and 
     /// the OptionalConsoleReadLine is used to read the first line.
     /// If a console is not used, then inReader.Peek() is called early instead.
-    /// </para><para>�</para>
+    /// </para><para>�</para>
     ///
     /// <para>Further lines are read using OptionalConsoleReadLine().
     /// If not provided, lines are read using inReader.ReadLine().</para>
-    /// <para>�</para>
+    /// <para>�</para>
```
Comment (dsyme) [inline src/fsharp/fsi/fsi.fsi]: What's this change? Code alignment? thanks

File: src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.fs (hunk 5 above)
Comment (ncave) [inline src/fsharp/FSharp.Compiler.Service.netcore/output/FSIstrings.fs]: Of course, that's there just temporary until I get the fssrgen working (the bootstrap fssrgen does not generate netcore-compatible code right now).

File: src/fsharp/fsi/fsi.fs (hunk 12 above)
Comment (ncave) [inline src/fsharp/fsi/fsi.fs]: yes, with ms/visualfsharp, just replacing Close with Dispose.

File: src/absil/il.fs (hunk 2 above)
Comment (ncave) [inline src/absil/il.fs]: System.Environment.OSVersion is unsupported on dotnetcore, that's why I used the ms/visualfsharp version here.

Comment (ncave) [inline src/absil/il.fs]: System.Environment.OSVersion is unsupported on dotnetcore, that's why I used the ms/visualfsharp version here.

File: src/absil/ilread.fs (hunk 4 above)
Comment (ncave) [inline src/absil/ilread.fs]: See src/absil/il.fs above for the reason to use ms/visualfsharp version.

File: src/fsharp/fsi/fsi.fs (hunk 10 above)
Comment (ncave) [inline src/fsharp/fsi/fsi.fs]: This is so all paths have the same return type.

Comment (ncave) [inline src/fsharp/fsi/fsi.fs]: This is just so all paths have the same return type.

File: src/fsharp/fsi/fsi.fs (hunk 11 above)
Comment (ncave) [inline src/fsharp/fsi/fsi.fs]: This is aligning with ms/visualfsharp.

**Comment by ncave** [comment]
Added the dotnet-fssrgen tool, it was the easiest way to be able to generate correct netcore versions.

**Comment by enricosada** [comment]
@ncave the 73fa783 is ok, but we can use the `dotnet-fssrgen` as .net cli tool (that's the expected usage).
I'll check and send a comment/pr soon

**Comment by enricosada** [comment]
@ncave ok publish `dotnet-fssrgen` version `3.0.1`

you can add this to project.json (ref [test](https://github.com/fsprojects/FsSrGen/blob/a8e4b34a7b1e122546030b621b2f5d9e63566c7b/test/use-dotnet-fssrgen-as-tool/project.json#L14-L16) )

```json

    "tools": {
        "dotnet-fssrgen": { "version": "3.0.1" }
    },
```

so after `dotnet restore`, you can do `dotnet fssrgen` with same arguments, instead of `dotnet-fssrgen`

File: build.fsx [hunk 14]
```diff
@@ -214,6 +214,49 @@ Target "GitHubRelease" (fun _ ->
...
+    Shell.Exec(fsYacc, @"../pppars.fsy" + lexArgs + yaccArgs + module3 + open3 + " -o pppars.fs", outPath) |> assertExitCodeZero
+)
+
+Target "DotnetCliBuild" (fun _ ->
+    let netcoreFW = "netstandard1.5"
+    let options1 = sprintf " --framework %s --configuration Release" netcoreFW
+    let options2 = " --configuration Release"
+    let outPath = @"src/fsharp/FSharp.Compiler.Service.netcore/"
+    Shell.Exec("dotnet", "restore", outPath) |> ignore //assertExitCodeZero
+    Shell.Exec("dotnet", "build" + options1, outPath) |> assertExitCodeZero
```
Comment (enricosada) [inline build.fsx]: just do `pack`, `pack` already do `build` if not updated
also, you can use `--verbose` flag to write more diagnostics

```
Shell.Exec("dotnet", "--verbose pack" + options2, outPath)
```

File: build.fsx [hunk 15]
```diff
@@ -214,6 +214,49 @@ Target "GitHubRelease" (fun _ ->
 )
 
 // --------------------------------------------------------------------------------------
+// .NET CLI and .NET Core
+
+let isDotnetCLIInstalled = Shell.Exec("dotnet", "--version") = 0
```
Comment (enricosada) [inline build.fsx]: use `--info`, it's better because write some diagnostics info, instead `--version` only print the version

File: src/absil/il.fs (hunk 2 above)
Comment (dsyme) [inline src/absil/il.fs]: Hmmm.. basically we don't want to be throwing exceptions in ilwrite.fs in FCS (the only place this is used) if we can detect that we are on Windows, .NET Framework.  If you could find some way to avoid relying on the try/catch in ilwrite.fs when on .NET Framework that would be great. Thanks

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+    "../vs/SimpleServices.fsi",
+    "../vs/SimpleServices.fs",
+    "../fsi/fsi.fsi",
+    "../fsi/fsi.fs"
+  ],
+  "dependencies": {
+    "Microsoft.FSharp.Core.netcore": "1.0.0-*",
+    "Microsoft.NETCore.App": {
+      "type": "platform",
+      "version": "1.0.0-rc2-3002426"
```
Comment (enricosada) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: `Microsoft.NETCore.App` is for console app ( `compilationOptions.emitEntryPoint` = `true` )

A library should use netstandard.library ( [info](https://github.com/enricosada/fsharp-dotnet-cli-samples/wiki/Migrate-fsproj-To-.NET-CLI#add-net-core-framework) )

```
"NETStandard.Library": "1.5.0-rc2-23925"
```

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+    }
+  },
+  "tools": {
+    "dotnet-fssrgen": {
+      "version": "2.0.1",
```
Comment (enricosada) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: use version `3.0.1` and you can remove the `imports`

File: src/fsharp/fsi/fsi.fs (hunk 8 above)
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: I think undo this change (don't make referencedAssemblies a global)

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+      "imports": [
+        "dnxcore50",
+        "portable-net45+win81",
+        "netstandard1.3"
+      ]
+    },
+    "dotnet-compile-fsc": {
```
Comment (enricosada) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: that should not be there, are you using .net cli version `1.0.0-beta-002071`? if not pls use that version

File: src/fsharp/fsi/fsi.fs (hunk 10 above)
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: ok thanks

File: src/fsharp/fsi/fsi.fs (hunk 11 above)
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: Please prefer the FCS version, thanks

File: src/fsharp/fsi/fsi.fs (hunk 12 above)
Comment (dsyme) [inline src/fsharp/fsi/fsi.fs]: ok

File: src/fsharp/vs/Reactor.fs [hunk 19]
```diff
@@ -9,6 +9,14 @@ open Microsoft.FSharp.Control
 open Microsoft.FSharp.Compiler
 open Microsoft.FSharp.Compiler.Lib
 
+#if COMPILER_SERVICE_CORECLR
+    //TODO: Trace not enabled in System.Diagnostics, so just a stub for now
+    type internal Trace() =
+        static member TraceInformation(format: string, [<ParamArray>] args: obj []) =
```
Comment (dsyme) [inline src/fsharp/vs/Reactor.fs]: I think we want this marked as ``[<Conditional("TRACE")>]``

File: src/fsharp/vs/SimpleServices.fs [hunk 20]
```diff
@@ -285,10 +287,13 @@ namespace Microsoft.FSharp.Compiler.SimpleSourceCodeServices
             let assemblyBuilderRef = ref None
             let tcImportsCapture = Some (fun tcImports -> tcImportsRef := Some tcImports)
 
-            let debugInfo =  otherFlags |> Array.exists (fun arg -> arg = "-g" || arg = "--debug:+" || arg = "/debug:+")
             // Function to generate and store the results of compilation 
+#if FX_NO_APP_DOMAINS
+            let dynamicAssemblyCreator = None
+#else
+            let debugInfo =  otherFlags |> Array.exists (fun arg -> arg = "-g" || arg = "--debug:+" || arg = "/debug:+")
```
Comment (dsyme) [inline src/fsharp/vs/SimpleServices.fs]: The ``CompileToDynamicAssembly`` methods aren't going to work if ``dynamicAssemblyCreator`` is ``None``.  Maybe elide the methods from the .NET Core API if they aren't working?

**Comment by dsyme** [comment]
Good progress!

File: tools/dotnet-fssrgen/Program.fs [hunk 21]
```diff
@@ -0,0 +1,465 @@
+namespace FSSRGen
```
Comment (enricosada) [inline tools/dotnet-fssrgen/Program.fs]: remove that, see my https://github.com/fsharp/FSharp.Compiler.Service/pull/556#issuecomment-212895651

File: tools/dotnet-fssrgen/NuGet.Config [hunk 22]
```diff
@@ -0,0 +1,10 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (enricosada) [inline tools/dotnet-fssrgen/NuGet.Config]: remove that, see https://github.com/fsharp/FSharp.Compiler.Service/pull/556#issuecomment-212895651

File: tools/dotnet-fssrgen/dotnet-fssrgen.xproj [hunk 23]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (enricosada) [inline tools/dotnet-fssrgen/dotnet-fssrgen.xproj]: remove that file, see https://github.com/fsharp/FSharp.Compiler.Service/pull/556#issuecomment-212895651

File: tools/dotnet-fssrgen/project.json [hunk 24]
```diff
@@ -0,0 +1,39 @@
+{
```
Comment (enricosada) [inline tools/dotnet-fssrgen/project.json]: remove that file, see https://github.com/fsharp/FSharp.Compiler.Service/pull/556#issuecomment-212895651

File: src/fsharp/fsi/fsi.fs [hunk 25]
```diff
@@ -2694,7 +2724,7 @@ type FsiEvaluationSession (fsi: FsiEvaluationSessionHostConfig, argv:string[], i
     static member GetDefaultConfiguration(fsiObj:obj, useFsiAuxLib) = 
     
         let rec tryFindMember (name : string) (memberType : MemberTypes) (declaringType : Type) =
-            match declaringType.GetMember(name, memberType, BindingFlags.Instance ||| BindingFlags.Public ||| BindingFlags.NonPublic) with
+            match declaringType.GetMember(name) with //, memberType, BindingFlags.Instance ||| BindingFlags.Public ||| BindingFlags.NonPublic) with
```
Comment (enricosada) [inline src/fsharp/fsi/fsi.fs]: can you remove the code, instead of comment it out?
comment out code is bad on diff

File: build.fsx (hunk 14 above)
Comment (ncave) [inline build.fsx]: I'm not sure ---verbose is supported on the latest cli.

Comment (enricosada) [inline build.fsx]: i think it should, you need to pass `--verbose` before the command name.
so `dotnet --verbose build` not `dotnet build --verbose`
btw we should use .net cli version `1.0.0-beta-002071` until rc2 it's officially released.
because we need to pin to one specific version (latest may not always work with f# or c#) and `1.0.0-beta-002071` works ok (xplat too)

File: build.fsx [hunk 26]
```diff
@@ -232,28 +231,22 @@ Target "DotnetCliParse" (fun _ ->
     let open1 = @" --open Microsoft.FSharp.Compiler.AbstractIL"
     let open2 = @" --open Microsoft.FSharp.Compiler"
     let open3 = @" --open Microsoft.FSharp.Compiler"
-    let run1 = @"run ../../src/fsharp/FSComp.txt ../../" + outPath + "FSComp.fs ../../" + outPath + "FSComp.resx"
-    let run2 = @"run ../../src/fsharp/fsi/FSIstrings.txt ../../" + outPath + "FSIstrings.fs ../../" + outPath + "FSIstrings.resx"
+    let gen1 = @"../FSComp.txt ./FSComp.fs ./FSComp.resx"
+    let gen2 = @"../fsi/FSIstrings.txt ./FSIstrings.fs ./FSIstrings.resx"
+    let options = " --info --configuration Release"
```
Comment (enricosada) [inline build.fsx]: i dont think `--info` is supported by `pack`

File: build.fsx [hunk 27]
```diff
@@ -232,28 +231,22 @@ Target "DotnetCliParse" (fun _ ->
...
 
-Target "DotnetCliBuild" (fun _ ->
-    let netcoreFW = "netstandard1.5"
-    let options1 = sprintf " --framework %s --configuration Release" netcoreFW
-    let options2 = " --configuration Release"
-    let outPath = @"src/fsharp/FSharp.Compiler.Service.netcore/"
-    Shell.Exec("dotnet", "restore", outPath) |> ignore //assertExitCodeZero
-    Shell.Exec("dotnet", "build" + options1, outPath) |> assertExitCodeZero
-    Shell.Exec("dotnet", "pack" + options2, outPath) |> assertExitCodeZero
+    Shell.Exec("dotnet", "pack" + options, outPath) |> assertExitCodeZero
```
Comment (enricosada) [inline build.fsx]:
```fsharp
Shell.Exec("dotnet", "--verbose pack" + options, outPath) |> assertExitCodeZero
```

File: build.fsx [hunk 28]
```diff
@@ -214,6 +214,42 @@ Target "GitHubRelease" (fun _ ->
...
+    let yaccArgs = @" --internal --parslib Internal.Utilities.Text.Parsing"
+    let module1 = @" --module Microsoft.FSharp.Compiler.AbstractIL.Internal.AsciiParser"
+    let module2 = @" --module Microsoft.FSharp.Compiler.Parser"
+    let module3 = @" --module Microsoft.FSharp.Compiler.PPParser"
+    let open1 = @" --open Microsoft.FSharp.Compiler.AbstractIL"
+    let open2 = @" --open Microsoft.FSharp.Compiler"
+    let open3 = @" --open Microsoft.FSharp.Compiler"
+    let gen1 = @"../FSComp.txt ./FSComp.fs ./FSComp.resx"
+    let gen2 = @"../fsi/FSIstrings.txt ./FSIstrings.fs ./FSIstrings.resx"
+    let options = " --info --configuration Release"
```
Comment (enricosada) [inline build.fsx]: `--info` is not needed

File: build.fsx [hunk 29]
```diff
@@ -214,6 +214,42 @@ Target "GitHubRelease" (fun _ ->
...
+    Shell.Exec("dotnet fssrgen", gen1, outPath) |> assertExitCodeZero
+    Shell.Exec("dotnet fssrgen", gen2, outPath) |> assertExitCodeZero
+    Shell.Exec(fsLex, @"../lex.fsl --unicode" + lexArgs + " -o lex.fs", outPath) |> assertExitCodeZero
+    Shell.Exec(fsLex, @"../pplex.fsl --unicode" + lexArgs + " -o pplex.fs", outPath) |> assertExitCodeZero
+    Shell.Exec(fsLex, @"../../absil/illex.fsl --unicode" + lexArgs + " -o illex.fs", outPath) |> assertExitCodeZero
+    Shell.Exec(fsYacc, @"../../absil/ilpars.fsy" + lexArgs + yaccArgs + module1 + open1 + " -o ilpars.fs", outPath) |> assertExitCodeZero
+    Shell.Exec(fsYacc, @"../pars.fsy" + lexArgs + yaccArgs + module2 + open2 + " -o pars.fs", outPath) |> assertExitCodeZero
+    Shell.Exec(fsYacc, @"../pppars.fsy" + lexArgs + yaccArgs + module3 + open3 + " -o pppars.fs", outPath) |> assertExitCodeZero
+
+    Shell.Exec("dotnet", "pack" + options, outPath) |> assertExitCodeZero
```
Comment (enricosada) [inline build.fsx]: use `--verbose` (before command name)

```fsharp
Shell.Exec("dotnet", "--verbose pack" + options, outPath) |> assertExitCodeZero
```

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json [hunk 30]
```diff
@@ -0,0 +1,288 @@
...
+    "../fsi/fsi.fsi",
+    "../fsi/fsi.fs"
+  ],
+  "dependencies": {
+    "Microsoft.FSharp.Core.netcore": "1.0.0-*",
+    "NETStandard.Library": "1.5.0-rc2-23925"
+  },
+  "tools": {
+    "dotnet-fssrgen": "3.0.1",
+    "dotnet-compile-fsc": {
```
Comment (enricosada) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: `dotnet-compile-fsc` mean you use latest version of .net core, you should use version `1.0.0-beta-002071` ( see [docs](https://github.com/enricosada/fsharp-dotnet-cli-samples/wiki/Status) )

We need to pin to one specific version of .net cli (until -rc2 is released), because latest can be broken.

**Comment by enricosada** [comment]
great wip @ncave

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json (hunk 16 above)
Comment (ncave) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: I tried to switch to NETStandard.Library, but had to reverse that since I'm getting too many errors for missing stuff.

File: src/absil/ilread.fs (hunk 4 above)
Comment (ncave) [inline src/absil/ilread.fs]: try/catch removed as requested, MemoryMappedFiles should work everywhere anyway (maybe even on mono).

File: src/fsharp/FSharp.Compiler.Service.netcore/project.json (hunk 16 above)
Comment (enricosada) [inline src/fsharp/FSharp.Compiler.Service.netcore/project.json]: if you open `project.json.lock` and search for `Microsoft.NETCore.App`, you see it reference some packages. NETCore.App is an empty package, just reference other packages.
You can directly use the depednencies of NETCore.App inside the project.json, it's the same.
After that, we can remove unused dependencies (like csharp and vb) and others.

I think the real list is https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/FSharp.Compiler.netcore.nuget/Microsoft.FSharp.Compiler.netcore.nuspec
Fixing the -rc2 build should be enough

File: build.fsx [hunk 31]
```diff
@@ -214,6 +214,42 @@ Target "GitHubRelease" (fun _ ->
...
+    let module2 = @" --module Microsoft.FSharp.Compiler.Parser"
+    let module3 = @" --module Microsoft.FSharp.Compiler.PPParser"
+    let open1 = @" --open Microsoft.FSharp.Compiler.AbstractIL"
+    let open2 = @" --open Microsoft.FSharp.Compiler"
+    let open3 = @" --open Microsoft.FSharp.Compiler"
+    let gen1 = @" ../FSComp.txt ./FSComp.fs ./FSComp.resx"
+    let gen2 = @" ../fsi/FSIstrings.txt ./FSIstrings.fs ./FSIstrings.resx"
+    let options = " --configuration Release"
+    
+    Shell.Exec("dotnet", "restore", outPath) |> ignore //assertExitCodeZero
```
Comment (enricosada) [inline build.fsx]: `assertExitCodeZero` should not be commented out. if restore fails, we are not getting packages

File: build.fsx [hunk 32]
```diff
@@ -214,6 +214,42 @@ Target "GitHubRelease" (fun _ ->
...
+
+    Shell.Exec("dotnet", "fssrgen" + gen1, outPath) |> assertExitCodeZero
+    Shell.Exec("dotnet", "fssrgen" + gen2, outPath) |> assertExitCodeZero
```
Comment (enricosada) [inline build.fsx]: maybe inline `gen1`, `gen2`, i think

```
Shell.Exec("dotnet", "fssrgen ../FSComp.txt ./FSComp.fs ./FSComp.resx", outPath) |> assertExitCodeZero
```

it's easier to read, instead of try find the variables

File: src/absil/ilread.fs (hunk 4 above)
Comment (enricosada) [inline src/absil/ilread.fs]: memory mapped should work only on windows. They dont use the memorymapped bcl classes, instead use windows kernel api ( ref [source](https://github.com/fsharp/FSharp.Compiler.Service/blob/e3a3266bac7ec56f6de8b79a283c311ccbc99a56/src/absil/ilread.fs#L119) )
That a nice refactoring/improvement to do, move current implementation to [bcl MemoryMapped  .net classes](https://msdn.microsoft.com/en-us/library/dd997372(v=vs.110).aspx)

Comment (enricosada) [inline src/absil/ilread.fs]: added https://github.com/Microsoft/visualfsharp/issues/1094 to track the kernel api vs bcl issue

File: build.fsx [hunk 33]
```diff
@@ -214,6 +214,42 @@ Target "GitHubRelease" (fun _ ->
 )
 
 // --------------------------------------------------------------------------------------
+// .NET CLI and .NET Core
+
+let isDotnetCLIInstalled = Shell.Exec("dotnet", "--info") = 0
```
Comment (dsyme) [inline build.fsx]: I think this line is crashing in AppVeyor CI

```
[00:01:46] Error: System.ComponentModel.Win32Exception (0x80004005): The system cannot find the file specified
[00:01:46] 	   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
[00:01:46] 	   at System.Diagnostics.Process.Start()
```

Note I think we could accept the PR if it is green (i.e. even if CoreCLR support is not running in CI - we can enable that next)

Comment (enricosada) [inline build.fsx]: Atm is expected. Dotnet cli is not installed. 
I'll try to add appveyor/travis

Comment (enricosada) [inline build.fsx]: Nvm, it should not throw

File: src/fsharp/fsi/fsi.fs (hunk 11 above)
Comment (ncave) [inline src/fsharp/fsi/fsi.fs]: That's just slight reordering in order to avoid AppDomain for netcore, otherwise functionality is fully retained.

File: src/absil/il.fs (hunk 2 above)
Comment (ncave) [inline src/absil/il.fs]: try/catch was removed.

File: src/absil/ilread.fs (hunk 4 above)
Comment (ncave) [inline src/absil/ilread.fs]: Thanks for opening the [#1094](https://github.com/Microsoft/visualfsharp/issues/1094) issue, I incorrectly assumed it was using the bcl underneath.

Comment (enricosada) [inline src/absil/ilread.fs]: i found about that searching for perfomance improvement. 
because the function are not inlined, are called a lot (really, like getByte), and is a bit of tax. 
i think if we can inline that's going away too, but i found really hard to do that because there was also binaryfile.
maybe i need to continue that old branch sometimes. or someone from community can volounteer (it's going to help a lot in .net core and mono)

File: build.fsx (hunk 33 above)
Comment (ncave) [inline build.fsx]: As a temporary measure I disabled it for anything but local builds.

**Comment by ncave** [comment]
@dsyme tests are passing, I tried to address most of the review comments, please let me know what you still think needs to to be changed

File: src/absil/ilread.fs [hunk 34]
```diff
@@ -3970,7 +3970,8 @@ let ClosePdbReader pdb =
...
+
+    if not runningOnMono then
```
Comment (dsyme) [inline src/absil/ilread.fs]: It's not a blocker, but just to note that ``runningOnMono`` is not the opposite of ``runningOnWindows`` - you can run Mono on Windows, for example (though very few people do these days).

**Comment by dsyme** [comment]
OK, merging, great work!!!

**Comment by dsyme** [comment]
I adjusted the build.fsx to allow local builds without CoreCLR installed here: https://github.com/fsharp/FSharp.Compiler.Service/commit/951b60d4aa573f78af3696bfd4e00ab9eda26031

**Comment by ncave** [comment]
Awesome, thanks!
